obs-studio: 28.0.3 (additional to 27.x)

SebTM opened the pull request:
###### Description of changes

Depends on #194572 #194518 (✅ merged)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

SebTM committed:
obs-studio: add 28.0.3 (parallel to 27.x)

SebTM commented:
```
Attempted path: share/obs/obs-studio/locale/en-US.ini
Attempted path: /nix/store/kdpiplmmvdgw3yync5wwhp8akkynghnf-obs-studio-28.0.3/share/obs/obs-studio/locale/en-US.ini
Attempted path: share/obs/obs-studio/locale.ini
Attempted path: /nix/store/kdpiplmmvdgw3yync5wwhp8akkynghnf-obs-studio-28.0.3/share/obs/obs-studio/locale.ini
Attempted path: share/obs/obs-studio/themes/Yami.qss
Attempted path: /nix/store/kdpiplmmvdgw3yync5wwhp8akkynghnf-obs-studio-28.0.3/share/obs/obs-studio/themes/Yami.qss
info: Using EGL/X11
info: CPU Name: AMD Ryzen 7 PRO 4750U with Radeon Graphics
info: CPU Speed: 1700.000MHz
info: Physical Cores: 8, Logical Cores: 16
info: Physical Memory: 31329MB Total, 13964MB Free
info: Kernel Version: Linux 5.15.71
info: Distribution: NixOS "22.11"
info: Session Type: x11
info: Window System: X11.0, Vendor: The X.Org Foundation, Version: 1.20.14
info: Qt Version: 5.15.5 (runtime), 5.15.5 (compiled)
info: Portable mode: false
Attempted path: share/obs/obs-studio/themes/Dark/no_sources.svg
Attempted path: /nix/store/kdpiplmmvdgw3yync5wwhp8akkynghnf-obs-studio-28.0.3/share/obs/obs-studio/themes/Dark/no_sources.svg
QMetaObject::connectSlotsByName: No matching signal for on_tbar_position_valueChanged(int)
QMetaObject::connectSlotsByName: No matching signal for on_actionShowTransitionProperties_triggered()
QMetaObject::connectSlotsByName: No matching signal for on_actionHideTransitionProperties_triggered()
info: OBS 28.0.3 (linux)
info: ---------------------------------
info: ---------------------------------
info: audio settings reset:
	samples per sec: 48000
	speakers:        2
	max buffering:   960 milliseconds
	buffering type:  dynamically increasing
info: ---------------------------------
info: Initializing OpenGL...
info: Loading up OpenGL on adapter AMD AMD RENOIR (LLVM 14.0.6, DRM 3.42, 5.15.71)
info: OpenGL loaded successfully, version 4.6 (Core Profile) Mesa 22.1.7, shading language 4.60
info: ---------------------------------
info: video settings reset:
	base resolution:   640x480
	output resolution: 640x480
	downscale filter:  Bicubic
	fps:               30/1
	format:            NV12
	YUV mode:          Rec. 709/Partial
info: NV12 texture support not available
info: P010 texture support not available
info: Audio monitoring device:
	name: Default
	id: default
info: ---------------------------------
info: ---------------------------------
info:   Loaded Modules:
info: ---------------------------------
info: ==== Startup complete ===============================================
**error: Encoder ID 'obs_x264' not found
error: Failed to get properties for encoder '' (ffmpeg_aac)
error: Could not enumerate any AAC encoder bitrates
error: Failed to create aac streaming encoder (simple output)**
info: Freeing OBS context data
warning: 1 views remain at shutdown
info: == Profiler Results =============================
info: run_program_init: 1022.73 ms
info:  ┣OBSApp::AppInit: 3.835 ms
info:  ┃ ┗OBSApp::InitLocale: 0.908 ms
info:  ┗OBSApp::OBSInit: 211.068 ms
info:    ┣obs_startup: 62.216 ms
info:    ┗OBSBasic::OBSInit: 115.947 ms
info:      ┣OBSBasic::InitBasicConfig: 0.134 ms
info:      ┣OBSBasic::ResetAudio: 0.337 ms
info:      ┣OBSBasic::ResetVideo: 114.975 ms
info:      ┣OBSBasic::InitOBSCallbacks: 0.012 ms
info:      ┣OBSBasic::InitHotkeys: 0.041 ms
info:      ┣obs_load_all_modules2: 0.024 ms
info:      ┗OBSBasic::ResetOutputs: 0.132 ms
info: obs_hotkey_thread(25 ms): min=0.348 ms, median=0.943 ms, max=25.107 ms, 99th percentile=25.107 ms, 96.2963% below 25 ms
info: audio_thread(Audio): min=0.017 ms, median=0.025 ms, max=0.04 ms, 99th percentile=0.04 ms
info: obs_graphics_thread(33.3333 ms): min=0.312 ms, median=0.413 ms, max=2.573 ms, 99th percentile=2.573 ms, 100% below 33.333 ms
info:  ┣tick_sources: min=0.001 ms, median=0.002 ms, max=0.003 ms, 99th percentile=0.003 ms
info:  ┣output_frame: min=0.251 ms, median=0.333 ms, max=2.521 ms, 99th percentile=2.521 ms
info:  ┃ ┗gs_context(video->graphics): min=0.246 ms, median=0.329 ms, max=2.518 ms, 99th percentile=2.518 ms
info:  ┃   ┣render_video: min=0.024 ms, median=0.039 ms, max=0.697 ms, 99th percentile=0.697 ms
info:  ┃   ┃ ┗render_main_texture: min=0.017 ms, median=0.029 ms, max=0.687 ms, 99th percentile=0.687 ms
info:  ┃   ┗gs_flush: min=0.006 ms, median=0.01 ms, max=0.018 ms, 99th percentile=0.018 ms
info:  ┗render_displays: min=0.015 ms, median=0.026 ms, max=0.063 ms, 99th percentile=0.063 ms
info: =================================================
info: == Profiler Time Between Calls ==================
info: obs_hotkey_thread(25 ms): min=25.434 ms, median=26.249 ms, max=49.981 ms, 7.69231% within ±2% of 25 ms (0% lower, 92.3077% higher)
info: obs_graphics_thread(33.3333 ms): min=32.577 ms, median=33.343 ms, max=33.926 ms, 95.2381% within ±2% of 33.333 ms (4.7619% lower, 0% higher)
info: =================================================
info: Number of memory leaks: 304
```

Builds now successfully but does not start with this error: `Failed to create aac streaming encoder (simple output)`

BenediktSeidl commented:
I think the problem is, that the plugin folder is cleared. At least i got it somehow working with the following hackish diff (which breaks the nixos plugin system for obs, it's only for demonstration!)

```diff
index 5303991f1fb..6db7c5239e7 100644
--- a/pkgs/applications/video/obs-studio/wrapper.nix
+++ b/pkgs/applications/video/obs-studio/wrapper.nix
@@ -22,16 +22,16 @@ symlinkJoin {
       wrapCommandLine = [
           "wrapProgram"
           "$out/bin/obs"
-          ''--set OBS_PLUGINS_PATH "${pluginsJoined}/lib/obs-plugins"''
-          ''--set OBS_PLUGINS_DATA_PATH "${pluginsJoined}/share/obs/obs-plugins"''
+          ''--set OBS_PLUGINS_PATH "${obs-studio}/lib/obs-plugins"''
+          ''--set OBS_PLUGINS_DATA_PATH "${obs-studio}/share/obs/obs-plugins"''
         ] ++ pluginArguments;
     in ''
     ${concatStringsSep " " wrapCommandLine}
 
     # Remove unused obs-plugins dir to not cause confusion
-    rm -r $out/share/obs/obs-plugins
+    # rm -r $out/share/obs/obs-plugins
     # Leave some breadcrumbs
-    echo 'Plugins are at ${pluginsJoined}/share/obs/obs-plugins' > $out/share/obs/obs-plugins-README
+    # echo 'Plugins are at ${pluginsJoined}/share/obs/obs-plugins' > $out/share/obs/obs-plugins-README
   '';
 
   inherit (obs-studio) meta;
```

Please also note that with obs version 28, there is no need for the external websocket plugin. i did not manage to build version 5.0.2 of it, but it's also not needed, because it is now included in obs (see https://github.com/obsproject/obs-studio/releases/tag/28.0.0): 
> Added obs-websocket 5.0 as a first-party plugin [tt2468]

So i think `obs-studio-plugins.obs-websocket` should to be removed with/after the update to version 28 of obs.

Sadly my nixos skills are not good enough to really help, but i hope this information saves you some time.

SebTM commented:
Thanks for your research and pointing this out! It seems, that we can not continue to drop the shipped plugins like before (?), maybe we can do something like splitting the plugin-files from the obs-build to include them in the combined-plugins output. But for now I have no idea how to archive this.

For the websocket-plugin: I also ran into this but forget to add it here as I dropped the plugins at some point for now until OBS28 runs (somehow).

raphaelr commented:
obs doesn't pick up its built-in plugins, setting OBS_PLUGINS_PATH to ./result/lib/obs-plugins (same with OBS_PLUGINS_DATA_PATH and share) makes startup succeed. Trying to figure out how to properly fix that.

Also we should consider building it with [Qt 6](https://github.com/obsproject/obs-studio/discussions/6481), which may cause plugin incompatibilities however.

SebTM commented:
> Also we should consider building it with [Qt 6](https://github.com/obsproject/obs-studio/discussions/6481), which may cause plugin incompatibilities however.

Checked the used dependencies, we missing `qtx11extras` for Qt6. As I didn't found libsForQt6 I would have imported qt6 and use `qt6.qtbase` and `qt6.qtsvg`? Or do we have to move the obs-studio package to `qt6-packages.nix`?

@raphaelr Thanks for looking into and helping out 🥇

BenediktSeidl commented:
i also wanted to share my latest success: i got obs working with an external plugin: https://github.com/BenediktSeidl/nixpkgs/commit/b4c7471f3825e27e9960298e6a5ced0b5de80f2e but the non wrapped version of obs will exit with an error, you described above. also `obs-backgroundremoval` can only be compiled from master.

SebTM commented:
Updated PR to build with qt6, added temporary fix (https://github.com/NixOS/nixpkgs/pull/194578#issuecomment-1268965958) to be able to run output for testing

miangraham commented:
Raphael has the problem right. obs-nix.c expects a relative OBS_PLUGIN_DESTINATION at runtime, but we need it to be absolute at build time.

I'd consider something like this: https://github.com/miangraham/nixpkgs/commit/286080568c51f5f3d3e3a06a1c8322bf15c98cd5

Leave OBS_PLUGIN_DESTINATION unchanged on the build side, but give the runtime side what it expects by pulling the prefix out of the string passed from cmake into obsconfig.h's #define. Slightly bigger change than one-line patching obs-nix.c, but future-proofs against other C code additions with similar expectations.

New variable used just for this at build time, as seen from cmake:
-- OBS_PLUGIN_DESTINATION_RELATIVE="lib/obs-plugins"

I've built but not runtime tested the above because my wayland system dumps core from QT before reaching the affected areas, which will be another bug to fix at some point. Should be testable from X.

miangraham commented:
Re: qt6, unrelated to obs-plugins, here's my wayland error for reference.

```console
$ nix-shell /home/ian/.cache/nixpkgs-review/rev-286080568c51f5f3d3e3a06a1c8322bf15c98cd5/shell.nix

.cache/nixpkgs-review/rev-286080568c51f5f3d3e3a06a1c8322bf15c98cd5 via ❄️  impure (review-shell)
$ obs --verbose
qt.qpa.plugin: Could not find the Qt platform plugin "wayland" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: offscreen, minimal, eglfs, linuxfb, minimalegl, xcb, vkkhrdisplay, vnc.

Aborted (core dumped)
```

raphaelr commented:
> I'd consider something like this: [miangraham@2860805](https://github.com/miangraham/nixpkgs/commit/286080568c51f5f3d3e3a06a1c8322bf15c98cd5)

Thanks!

> Re: qt6, unrelated to obs-plugins, here's my wayland error for reference

adding `qtwayland` to `buildInputs` fixes that - I guess that's not included by default in qt6 anymore?

miangraham commented:
Good call. This addition got me running on wayland (unwrapped): https://github.com/miangraham/nixpkgs/commit/e36a42b0c8fe327c0a722eb6135b2c8c66e2a635

```console
.cache/nixpkgs-review/rev-e36a42b0c8fe327c0a722eb6135b2c8c66e2a635 via ❄️  impure (review-shell)
λ obs --verbose
...
info: Platform: Wayland
...
info: Qt Version: 6.4.0 (runtime), 6.4.0 (compiled)
...
info:   Loaded Modules:
info:     vlc-video.so
info:     text-freetype2.so
info:     rtmp-services.so
info:     obs-x264.so
info:     obs-websocket.so
info:     obs-vst.so
info:     obs-transitions.so
info:     obs-outputs.so
info:     obs-filters.so
info:     obs-ffmpeg.so
info:     obs-browser.so
info:     linux-v4l2.so
info:     linux-pulseaudio.so
info:     linux-pipewire.so
info:     linux-capture.so
info:     linux-alsa.so
info:     image-source.so
info:     frontend-tools.so
info:     decklink-output-ui.so
info:     decklink-captions.so
info: ---------------------------------
info: ==== Startup complete ===============================================
...
info: =================================================
info: Number of memory leaks: 0
```

Lots of plugin build breakage as expected, but otherwise looks much better than I would have guessed this early.

Assuming the above works out, it should be safe to revert the changes to wrapper.nix and nuke obs-websocket.nix.

SebTM commented:
Thanks for digging into @miangraham I've encountered your feedback/patch to the PR to then try to get some plugins working.
@raphaelr I've added `qtwayland` to `buildInputs` or do we need a condition here?

Can I just remove obs-websocket plugin from nixpkgs to drop it or is there some kind of option where we should do this depending on obs-studio version for backwards-compatibility? Should this be added to this PR or a separate referenced here?

SebTM commented:
Successfully fixed build for `obs-pipewire-audio-capture` and working test-recording.

raphaelr commented:
> I've added `qtwayland` to `buildInputs` or do we need a condition here?

Just add it to buildInputs unconditionally.

> Can I just remove obs-websocket plugin from nixpkgs to drop it

Remove the file and replace the entry in plugins.nix with a throw: `throw "obs-websocket has been removed: Functionality has been integrated into obs-studio itself."`, so that users know what happened.

In any case, removing plugins is a breaking change and therefore requires an entry in the [release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2211.section.md#backward-incompatibilities-sec-release-2211-incompatibilities).

raphaelr commented on pkgs/applications/video/obs-studio/default.nix:
```diff
@@ -46,19 +51,20 @@ let
   };
 
   patches = [
     # Lets obs-browser build against CEF 90.1.0+
```
Do we still need this?

SebTM replied:
The code-part patched is consistent since 3 years, obs still builds without but I think there was a need for it even I can explain/have a scenario to test it. How should we handle this?

miangraham replied:
I don't see any obvious incompatibilities and it's better security. If there's no breakage observed I think it's better to keep it.

SebTM commented:
> > I've added `qtwayland` to `buildInputs` or do we need a condition here?
> 
> Just add it to buildInputs unconditionally.

✅ 

> 
> > Can I just remove obs-websocket plugin from nixpkgs to drop it
> 
> Remove the file and replace the entry in plugins/default.nix with a throw: `throw "obs-websocket has been removed: Functionality has been integrated into obs-studio itself."`, so that users know what happened.
> 
> In any case, removing plugins is a breaking change and therefore requires an entry in the [release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2211.section.md#backward-incompatibilities-sec-release-2211-incompatibilities).

Can I do it with this PR as the plugin will be removed with the upgrade to OBS28 or should it be another depending PR (like `librist` and `libajantv2`)

> 
> > -DENABLE_JACK=ON
> 
> Should add this to cmakeFlags, `libjack2` is already included in buildInputs, so it makes sense to enable it

Will add it later and check if the patches are still necessary.

miangraham commented:
Result of `nixpkgs-review pr 194578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-backgroundremoval</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-pipewire-audio-capture</li>
    <li>obs-studio-plugins.obs-websocket</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>libajantv2</li>
    <li>librist</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-vkcapture</li>
    <li>obs-studio-plugins.wlrobs</li>
  </ul>
</details>

miangraham commented:
Result of `nixpkgs-review pr 194578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-backgroundremoval</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-pipewire-audio-capture</li>
    <li>obs-studio-plugins.obs-websocket</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>libajantv2</li>
    <li>librist</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-vkcapture</li>
    <li>obs-studio-plugins.wlrobs</li>
  </ul>
</details>

SebTM commented:
Plugins already fixed for OBS28:
* obs-studio-plugins.obs-move-transition - https://github.com/NixOS/nixpkgs/pull/195051
* obs-studio-plugins.obs-pipewire-audio-capture - https://github.com/NixOS/nixpkgs/pull/195024

SebTM commented:
Updated PR to add `-DENABLE_JACK=ON`. When it's ready I will remove the picked commit for the required new packages.

ymatsiuk commented on pkgs/applications/video/obs-studio/default.nix:
```diff
@@ -46,19 +51,20 @@ let
   patches = [
     # Lets obs-browser build against CEF 90.1.0+
     ./Enable-file-access-and-universal-access-for-file-URL.patch
+    ./Provide-runtime-plugin-destination-as-relative-path.patch
```
Does it make more sense to set the variable we're patching in the wrapper (like we set some runtime dependencies in the preFixup phase)? 🤔

miangraham replied:
It would be much cleaner if we could set it there but I don't think it'll work. If I understand qtWrapperArgs properly that set just gets turned into an envvar, but OBS_PLUGIN_DESTINATION isn't a runtime envvar, it's just a cmake variable built into a #define.

See: [Injection from cmake to C](https://github.com/obsproject/obs-studio/blob/277dafab98c087c71deb005cbfaf9ea05be2efc8/libobs/obsconfig.h.in#L16), [Usage from C](https://github.com/obsproject/obs-studio/blob/277dafab98c087c71deb005cbfaf9ea05be2efc8/libobs/obs-nix.c#L57)

SebTM replied:
> OBS_PLUGIN_DESTINATION isn't a runtime envvar

Tried it out and didn't worked for me so this explains it 👍🏻

miangraham commented:
Result of `nixpkgs-review pr 194578` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-backgroundremoval</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-ndi</li>
    <li>obs-studio-plugins.obs-pipewire-audio-capture</li>
    <li>obs-studio-plugins.obs-websocket</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>obs-studio</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-vkcapture</li>
    <li>obs-studio-plugins.wlrobs</li>
  </ul>
</details>

miangraham commented:
> Updated PR to add `-DENABLE_JACK=ON`

Tested on https://github.com/NixOS/nixpkgs/commit/31b147d1613df86466bc5bf0cca86e0fd990df5f:

```console
info: ---------------------------------
info:   Loaded Modules:
...
info:     linux-jack.so
...
info: ---------------------------------

info: ------------------------------------------------
info: Loaded scenes:
info: - scene 'Main':
info:     - source: 'obs music in' (jack_output_capture)
info:         - monitoring: monitor and output
info:     - source: 'obs mic in' (jack_output_capture)
info:         - monitoring: monitor and output
...
info: ------------------------------------------------
```

It works! 🎉 

Qt is still sad:

```console
info: Qt Version: 5.15.5 (runtime), 5.15.5 (compiled)
```

Current PR is missing at least the `qt6Packages.callPackage` change from a [previously working commit](https://github.com/NixOS/nixpkgs/commit/f975db709dffadd1bb4ad6b8e3a0a5d3325188c2#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22R30423).

SebTM commented:
Updated PR to build against Qt6 again, removed cherry-picked dependencies as merged now.

raphaelr commented:
> Can I do it within this PR (in an additional commit) as the plugin will be removed with the upgrade to OBS28 or should it be another depending PR (like `librist` and `libajantv2`)

Do it in this PR; if you do it in a new PR master becomes broken (no obs-websocket but also no OBS 28).

blitz commented:
Great to see a new OBS coming in. :+1:  Can you rebase this PR? Otherwise, the build runs into:

```
❯ nix-build -A pkgs.obs-studio
error: Function called without required argument "librist" at /home/julian/src/own/nixpkgs/pkgs/applications/video/obs-studio/default.nix:40, did you mean "librest", "libXtst" or "libast"?
```

blitz commented:
I've tested it with some light recording workloads and it looks good. Fiddling with a V4L2 source made by Gnome Session crash, but that doesn't seem like it should be OBS's fault.

The Gnome Shell crash is this one and it's supposedly fixed in GNOME 43: https://gitlab.gnome.org/GNOME/mutter/-/issues/2433 (I'm running 42.4)

SebTM commented:
> > Can I do it within this PR (in an additional commit) as the plugin will be removed with the upgrade to OBS28 or should it be another depending PR (like `librist` and `libajantv2`)
> 
> Do it in this PR; if you do it in a new PR master becomes broken (no obs-websocket but also no OBS 28).

Done in separate commit within this PR

> Great to see a new OBS coming in. +1 Can you rebase this PR? Otherwise, the build runs into:

Rebased on current master

raphaelr commented:
obs-studio looks good; plugin status (nix build .#obs-studio-plugins.*name*):

https://obsproject.com/kb/obs-studio-28-plugin-compatibility

- [ ] looking-glass-obs: builds, do not have the hardware to test
- [ ] obs-backgroundremoval: build failure; needs building from master as indicated by BenediktSeidl
- [ ] obs-gstreamer: builds, need to test
- [ ] obs-hyperion: https://github.com/NixOS/nixpkgs/pull/190721
- [ ] obs-move-transition: https://github.com/NixOS/nixpkgs/pull/195051
- [ ] obs-multi-rtmp: build failure
- [ ] obs-ndi: unfree stuff, cannot even attempt to build, compatibility page says "in progress"
- [ ] obs-nvfbc: build failure
- [ ] obs-pipewire-audio-capture: https://github.com/NixOS/nixpkgs/pull/195024
- [ ] obs-vkcapture: builds, need to test
- [x] obs-websocket: removed
- [ ] wlrobs: https://github.com/NixOS/nixpkgs/pull/195029

This is just an overview of the current status, not a "this MUST be fixed" list.

miangraham commented:
obs-ndi non-compat tracking issue: https://github.com/Palakis/obs-ndi/issues/745

Upstream is broken against v28 until they finalize a fix.

Bleeding edge packagers like Alpine-edge, Arch, RPMFusion seem to have rolled forward without it. It's also got users in practice--see activity on the issue above for an idea.

Assuming we don't plan to wait for it, flag as broken until they have a new release and consider user impact as a factor in merging before or after 22.11.

Atemu commented:
Since many plugins aren't ported yet and the changes to the derivation seem fairly minimal, could we keep around both versions?

SebTM commented:
> Since many plugins aren't ported yet and the changes to the derivation seem fairly minimal, could we keep around both versions?

Sounds like a good idea, will try to have a look at it but mighty take some days (maybe the weekend).

SebTM commented:
> Since many plugins aren't ported yet and the changes to the derivation seem fairly minimal, could we keep around both versions?

I've had quiet a look today but as there are fundamental changes like `stdenv.mkDerivation` vs `mkDerivation` my solution would be two separate derivations even with some duplicate code as we intend to drop 27 in the near future? Otherwise I would need some help/ideas on a more generic approach/example

Atemu commented:
I'm not an expert in this but it seems `mkDerivation` as a direct parameter is a remnant of an older style of packaging QT apps. Just use `stdenv.mkDerivation` and add `wrapQtAppsHook` to `nativeBuildInputs` for both versions.

As for version selection, I'd use a pattern similar to Emacs. A generic.nix with a function that takes version and hash and returns the existing package function. Then add versioned files that simply call generic.nix with a certain version. (You could do the same in a single file, that's preference.)

In the derivation we can then check against the version for optional dependencies etc. using `finalAttrs.version` (or rec but it kinda sucks).

SebTM commented:
Thanks @Atemu again for your help - I'm not quite sure if it's fully the solution you proposed but there were some issues with importing additional packages in 27/28.nix and appending to buildInputs so I came across this solution in a library and got it adapted and working. Both versions 27.x and 28.x are building and possible to launch without errors.

One thing I'm unsure about, how to handle/proceed with the plugins as there are sometimes different build/things for LibObs etc. needed.

fpletz reviewed (approved):
Since we will have both versions available in nixpkgs, let's merge this.
